cpp/tests/lead_lag/lead_lag_test.cpp

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <algorithm>
+#include <cudf/aggregation.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <cudf/null_mask.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/type_lists.hpp>
+#include <functional>
+#include <initializer_list>
+#include <iterator>
+#include <memory>
+#include <rmm/device_buffer.hpp>
+
+using cudf::size_type;
+using namespace cudf::test;
+
+struct LeadLagWindowTest : public cudf::test::BaseFixture {
+};
+
+template <typename T>
+struct TypedLeadLagWindowTest : public cudf::test::BaseFixture {
+};
+
+using TypesForTest = cudf::test::Concat<cudf::test::IntegralTypes,
+                                        cudf::test::FloatingPointTypes,
+                                        cudf::test::DurationTypes,
+                                        cudf::test::TimestampTypes>;
+
+TYPED_TEST_CASE(TypedLeadLagWindowTest, TypesForTest);
+
+TYPED_TEST(TypedLeadLagWindowTest, LeadLagBasics)
+{
+  using T = int32_t;
+
+  auto const input_col =
+    fixed_width_column_wrapper<T>{0, 1, 2, 3, 4, 5, 0, 10, 20, 30, 40, 50}.release();
+  auto const input_size   = input_col->size();
+  auto const grouping_key = fixed_width_column_wrapper<int32_t>{0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1};
+  auto const grouping_keys = cudf::table_view{std::vector<cudf::column_view>{grouping_key}};
+
+  auto lead_3_output_col = cudf::grouped_rolling_window(
+    grouping_keys, input_col->view(), 4, 3, 1, cudf::make_lead_aggregation(3));
```

Comment:
It would be easier if we know what is 4, 3,  and 1, so better represent them with variables.

Reply:
Agreed. I have introduced variables for the window-specs, on all tests.